File: Python/ByRound/1919/1919_D_01_Tree.py
# Problem: CF 1919 D - 01 Tree
# https://codeforces.com/contest/1919/problem/D

"""
Purpose: Determine if a given array of distances from root to leaves in DFS order
         can correspond to a valid complete binary tree where edges have weights 0 or 1.

Algorithms/Techniques: 
    - Stack-based validation of DFS order compatibility with binary tree structure
    - Monotonic stack to track the depth of the most recent leaf processed in the current path

Time Complexity: O(n) for each test case, where n is the number of leaves.
Space Complexity: O(n) for the stack storage in the worst case.

"""
import sys

sys.setrecursionlimit(2147483647)
MOD = 1000000007
input = sys.stdin.readline


def solve():
    n = int(input())
    a = list(map(int, input().split()))
    s = [a[0]]  # Initialize stack with the first element of array
    for i in range(1, n):
        # While the last element in stack is greater than current depth and
        # can be reduced by 1 to match current depth, pop it (backtrack in DFS)
        while len(s) > 1 and s[-1] > a[i] and s[-1] - 1 == s[-2]:
            s.pop()
        # If the last element in stack is less than the current depth, 
        # we must extend the path (push new depth)
        if s[-1] < a[i]:
            s.append(a[i])
        # If the last element in stack is one more than the current depth,
        # update the last element (i.e., move to sibling in DFS)
        elif s[-1] - 1 == a[i]:
            s[-1] = a[i]
        # If they are equal and the depth is 0, it's invalid since root cannot have a sibling
        elif s[-1] == a[i]:
            if a[i] == 0:
                print("NO")
                return
        # Otherwise, it's an invalid sequence
        else:
            print("NO")
            return
    # After processing all elements, prune the stack by removing redundant depths
    while len(s) > 1 and s[-1] - 1 == s[-2]:
        s.pop()
    # If we end up with exactly one depth of 0, the sequence is valid
    if len(s) == 1 and s[0] == 0:
        print("YES")
    else:
        print("NO")


t = int(input())
for _ in range(t):
    solve()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1919/1919_D_01_Tree.cpp
// Problem: CF 1919 D - 01 Tree
// https://codeforces.com/contest/1919/problem/D

/*
Purpose: Determine if a given array of distances from root to leaves in DFS order can be represented by a valid complete binary tree where each internal edge has weight 0 or 1.

Algorithm:
- The key insight is that in a valid tree, each leaf's distance from the root must obey certain constraints based on the tree structure.
- For a complete binary tree with n leaves, we simulate the process of building the tree and check the consistency of the input array.
- We use a monotonic stack approach to precompute for each element the largest previous value smaller than current element (pre array) and the smallest next value larger than current element (suf array).
- Then, for each leaf, we check that either the previous or the next element is one less than its distance, which ensures the path to the leaf is valid.

Time Complexity: O(n) per test case
Space Complexity: O(n) for arrays and stack usage

Techniques:
- Monotonic stack
- Preprocessing with stack to find nearest smaller elements
- Constraint validation based on tree structure properties
*/

#include <cstdio>
#include <utility>

#define N 200005
#define P 1000000007
#define INF 1061109567
using namespace std;
typedef long long ll;
typedef long double ld;
typedef pair<int, int> PII;
typedef unsigned long long ull;
char buf[1 << 20], *p1, *p2;
#define GC                                                                     \
  (p1 == p2 && (p2 = (p1 = buf) + fread(buf, 1, 1 << 20, stdin), p1 == p2)     \
       ? 0                                                                     \
       : *p1++)
inline ll read() {
  ll x = 0, f = 1;
  char ch = GC;
  while (ch < '0' || ch > '9') {
    if (ch == '-')
      f = -1;
    ch = GC;
  }
  while (ch >= '0' && ch <= '9') {
    x = (x << 1) + (x << 3) + (ch ^ 48);
    ch = GC;
  }
  return x * f;
}
int n, cnt0, a[N], pre[N], suf[N];
int R, q[N];
void Solve() {
  n = read();
  cnt0 = 0;
  for (int i = 1; i <= n; i++)
    a[i] = read(), cnt0 += !a[i];  // Count number of 0s
  if (cnt0 != 1) {  // Root must have exactly one path of weight 0
    printf("NO\n");
    return;
  }
  a[0] = a[n + 1] = -1;
  q[R = 1] = 0;
  for (int i = 1; i <= n; i++) {
    while (R && a[q[R]] >= a[i])  // Maintain decreasing order in stack
      R--;
    pre[i] = a[q[R]];  // Store previous smaller element
    q[++R] = i;
  }
  q[R = 1] = n + 1;
  for (int i = n; i >= 1; i--) {
    while (R && a[q[R]] >= a[i])
      R--;
    suf[i] = a[q[R]];  // Store next smaller element
    q[++R] = i;
  }
  for (int i = 1; i <= n; i++)
    if (pre[i] != a[i] - 1 && suf[i] != a[i] - 1) {  // Check constraints
      printf("NO\n");
      return;
    }
  printf("YES\n");
}
int main() {
  for (int T = read(); T--;)
    Solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/